### PR TITLE
Heat capacity

### PR DESCRIPTION
**Pull request by d-rams**
This PR adds a simple heat capacity metric.  I have not included the TAP3-based heat capacity here since it's not commensurate with this one yet...

### CODE REVIEW

**Comment by drckf** [comment]
We should update all of the docstrings for the functions in `metrics.py`. For example, the doctoring for the `ReconstructionError.update` method is:
```
        Update the estimate for the reconstruction error using a batch
        of observations and a batch of reconstructions.

        Notes:
            Changes norm and mean_square_error in place.

        Args:
            minibatch (tensor (num_samples, num_units))
            reconstructions (tensor (num_samples, num))
            kwargs: key word arguments
                not used, but helpful for looping through metric functions

        Returns:
            None
```
But, in reality, all of these functions take in a `MetricState` object.

I was confused at first when I was looking at the code in this PR because `MetricState` used to contain tensors of just visible units and the methods acted on these tensors. Now, `MetricState` contains `State` objects, which contain tensors for all of the layers. This is the right way to do it -- but I got confused because the docstrings no longer match with what the functions are actually doing.

File: paysage/metrics.py [hunk 1]
```diff
@@ -361,3 +361,83 @@ def value(self) -> float:
...
+        Args:
+            Just the model is used since this is not a stochastic metric
+
+        Returns:
+            None
+
+        """
+        self.norm += 1
+        self.heat_capacity += be.mean(be.square(update_args.amodel
+                                   .joint_energy(update_args.random_samples)))
```
Comment (drckf) [inline paysage/metrics.py]: I know we discussed this briefly, but I am not sure if this function should compute the variance using the `random_samples` or the `samples` (i.e., the fantasy particles). Have you tried it the other way?

**Comment by drckf** [comment]
Also, we should convert the `EnergyGap` and `EnergyZScore` methods to use the `joint_energy` instead of the `marginal_free_energy` because the former is well-defined for deep models whereas the latter is not really computable. In fact, we should probably deprecate the `marginal_free_energy` function, which may simplify the code a bit. 

I am gonna put this here for now, but this should probably be an issue for a future PR rather than something to be done here.

**Comment by d-rams** [comment]
So, @drckf , the heat capacity estimated from randomly seeded gibbs sampling seems to be more consistent and increases almost monotonically during training -- for both TAP and CD1

On the other hand the heat capacity estimated from 'samples' is more irregular

**Comment by d-rams** [comment]
Here's a sample training run with TAP from random initialized sampling:
```End of epoch 0: 
-ReconstructionError: 8.546091
-EnergyDistance: 3.958037
-HeatCapacity: 6788.965885
Epoch took 35.75 seconds

End of epoch 1: 
-ReconstructionError: 7.848588
-EnergyDistance: 1.492231
-HeatCapacity: 16787.120312
Epoch took 35.99 seconds

End of epoch 2: 
-ReconstructionError: 7.766338
-EnergyDistance: 3.241121
-HeatCapacity: 24921.048958
Epoch took 36.07 seconds

End of epoch 3: 
-ReconstructionError: 7.443050
-EnergyDistance: 1.324916
-HeatCapacity: 32711.237500
Epoch took 36.04 seconds

End of epoch 4: 
-ReconstructionError: 7.443297
-EnergyDistance: 1.655846
-HeatCapacity: 36596.464583
Epoch took 36.43 seconds

End of epoch 5: 
-ReconstructionError: 7.261841
-EnergyDistance: 1.549744
-HeatCapacity: 44163.539583
Epoch took 36.00 seconds

End of epoch 6: 
-ReconstructionError: 7.224242
-EnergyDistance: 1.302964
-HeatCapacity: 46493.431250
Epoch took 36.18 seconds

End of epoch 7: 
-ReconstructionError: 7.200926
-EnergyDistance: 1.721646
-HeatCapacity: 50920.579167
Epoch took 36.05 seconds

End of epoch 8: 
-ReconstructionError: 7.207658
-EnergyDistance: 1.071913
-HeatCapacity: 55787.504167
Epoch took 36.32 seconds

End of epoch 9: 
-ReconstructionError: 7.208421
-EnergyDistance: 1.902518
-HeatCapacity: 57973.641667
Epoch took 42.26 seconds

Final performance metrics:
-ReconstructionError: 7.213598
-EnergyDistance: 1.792917
-HeatCapacity: 57396.000000
```

**Comment by d-rams** [comment]
And one seeded from 'samples':
```
End of epoch 0: 
-ReconstructionError: 8.546091
-EnergyDistance: 3.826556
-HeatCapacity: 3685.408333
Epoch took 44.22 seconds

End of epoch 1: 
-ReconstructionError: 7.848588
-EnergyDistance: 1.472322
-HeatCapacity: 7349.908333
Epoch took 48.03 seconds

End of epoch 2: 
-ReconstructionError: 7.766338
-EnergyDistance: 3.247717
-HeatCapacity: 1871.416667
Epoch took 47.27 seconds

End of epoch 3: 
-ReconstructionError: 7.443050
-EnergyDistance: 1.342238
-HeatCapacity: 13731.295833
Epoch took 46.78 seconds

End of epoch 4: 
-ReconstructionError: 7.443297
-EnergyDistance: 1.643400
-HeatCapacity: 22611.241667
Epoch took 45.79 seconds

End of epoch 5: 
-ReconstructionError: 7.261841
-EnergyDistance: 1.580270
-HeatCapacity: 8607.708333
Epoch took 38.49 seconds

End of epoch 6: 
-ReconstructionError: 7.224242
-EnergyDistance: 1.267709
-HeatCapacity: 16939.441667
Epoch took 38.81 seconds

End of epoch 7: 
-ReconstructionError: 7.200926
-EnergyDistance: 1.644327
-HeatCapacity: 33306.695833
Epoch took 37.10 seconds

End of epoch 8: 
-ReconstructionError: 7.207658
-EnergyDistance: 1.146606
-HeatCapacity: 13616.983333
Epoch took 36.67 seconds

End of epoch 9: 
-ReconstructionError: 7.208421
-EnergyDistance: 1.788684
-HeatCapacity: 49385.587500
Epoch took 36.51 seconds

Final performance metrics:
-ReconstructionError: 7.213598
-EnergyDistance: 1.733209
-HeatCapacity: 49192.308333
```

**Comment by d-rams** [comment]
You get the same kind of behavior with CD-1

**Comment by d-rams** [comment]
It seems to me like the true sampling from the model distribution is more useful as a training metric.  Which is favorable theoretically imo.

**Comment by d-rams** [comment]
OK, I'm convinced that theoretically the heat capacity should be sampled using the fantasy particles, not simply random samples from the model's visible layer.  In this case using `random_samples` was actually using Gaussian random hidden units which is not appropriate.

**Pull request merged**